smsotp/smsotp-authenticator/org.wso2.carbon.identity.authenticator/src/main/java/org/wso2/carbon/identity/authenticator/SMSOTPAuthenticator.java

```diff
@@ -0,0 +1,298 @@
+
+import javax.net.ssl.HttpsURLConnection;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.System;
+import java.net.MalformedURLException;
+import java.net.ProtocolException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+
+/**
+ * Authenticator of SMSOTP
+ */
+public class SMSOTPAuthenticator extends AbstractApplicationAuthenticator implements FederatedApplicationAuthenticator {
+
+    private static Log log = LogFactory.getLog(SMSOTPAuthenticator.class);
+    AuthenticationContext authContext = new AuthenticationContext();
+    Map<String, String> newAuthenticatorProperties;
+    private String otpToken;
+    private String mobile;
+    private String smsUrl = "";
+    private String clientId = "";
+    private String clientSecret = "";
+    private String fullUrl = "";
+
+    /**
+     * Check whether the authentication or logout request can be handled by the authenticator
+     */
+    public boolean canHandle(HttpServletRequest request) {
+        if (log.isDebugEnabled()) {
+            log.debug("Inside SMSOTPAuthenticator canHandle method");
+        }
+        return StringUtils.isNotEmpty(request.getParameter(SMSOTPConstants.CODE));
+    }
+
+    /**
+     * initiate the authentication request
+     */
+    @Override
+    protected void initiateAuthenticationRequest(HttpServletRequest request,
+                                                 HttpServletResponse response, AuthenticationContext context)
+            throws AuthenticationFailedException {
+
+        String resourceName = SMSOTPConstants.PROPERTIES_FILE;
+        ClassLoader loader = Thread.currentThread().getContextClassLoader();
+        Properties prop = new Properties();
+
+        InputStream resourceStream = loader.getResourceAsStream(resourceName);
+        try {
+            prop.load(resourceStream);
+        } catch (IOException e) {
+            throw new AuthenticationFailedException("Can not find the file", e);
+        }
+
+        newAuthenticatorProperties = context
+                .getAuthenticatorProperties();
+        newAuthenticatorProperties.put("username", prop.getProperty("username"));
+        newAuthenticatorProperties.put("password", prop.getProperty("password"));
+        newAuthenticatorProperties.put("from", prop.getProperty("from"));
+        newAuthenticatorProperties.put("text", prop.getProperty("text"));
+        context.setAuthenticatorProperties(newAuthenticatorProperties);
+
+        OneTimePassword token = new OneTimePassword();
+        String secret = OneTimePassword.getRandomNumber(SMSOTPConstants.SECRET_KEY_LENGTH);
+        otpToken = token.generateToken(secret, "" + SMSOTPConstants.NUMBER_BASE, SMSOTPConstants.NUMBER_DIGIT);
+        Object myToken = otpToken;
+        authContext.setProperty(otpToken, myToken);
```

Comment:
Use a constance variable to store the token in authentication context. We can't retrieve the value since this is dynamic.